File: factorial.js
var ffi = require('ffi')
var ref = require('ref');
var StructType  = require('ref-struct');


const SmiModule = StructType();
const SmiModulePtr = ref.refType(SmiModule);

SmiModule.defineProperty('name', 'string');
SmiModule.defineProperty('path', 'string');
SmiModule.defineProperty('organization', 'string');
SmiModule.defineProperty('contactinfo', 'string');
SmiModule.defineProperty('description', 'string');
SmiModule.defineProperty('reference', 'string');
SmiModule.defineProperty('language', 'int');
SmiModule.defineProperty('conformance', 'int');


var libfactorial = ffi.Library(__dirname + '/libfactorial.dll', {
  'factorial': [ 'uint64', [ 'int' ] ],
  'GetSmiModule': [ SmiModulePtr, [] ]
})

var smimod = libfactorial.GetSmiModule();

console.log('Your output: ', smimod, smimod.language, smimod.conformance);

if (process.argv.length < 3) {
  console.log('Arguments: ' + process.argv[0] + ' ' + process.argv[1] + ' <max>')
  process.exit()
}

var output = libfactorial.factorial(parseInt(process.argv[2]))

console.log('Your output: ' + output);

File: lib/factorial.c
#include <stdint.h>

#if defined(WIN32) || defined(_WIN32)
#define EXPORT __declspec(dllexport)
#else
#define EXPORT
#endif

/* SmiModule -- the main structure of a module                               */
typedef struct SmiModule {
    char       *name;
    char       *path;
    char       *organization;
    char       *contactinfo;
    char       *description;
    char       *reference;
    int     language;
    int     conformance;
} SmiModule;

EXPORT uint64_t factorial(int max) {
  int i = max;
  uint64_t result = 1;

  while (i >= 2) {
    result *= i--;
  }

  return result;
}

EXPORT SmiModule* GetSmiModule() {
  struct SmiModule *retval = malloc(sizeof(struct SmiModule));
  
  retval->language = 1;
  retval->conformance = 2;
  
  return retval;
}